### PR TITLE
[SCSB-155] build with Numpy 2.0

### PR DESCRIPTION
**Pull request by zacharyburnett**
resolves [SCSB-155](https://jira.stsci.edu/browse/SCSB-155)

Numpy 2.0 is scheduled for release on June 19th. Due to the C ABI updates, Python projects that build C extensions are required to build against Numpy 2.0 in order to use Numpy 2.0 in runtime.
This is backwards-compatible with Numpy versions in runtime as far back as `1.26`

this PR pins `numpy>=2.0.0rc2` in `build-system.requires`
> [!NOTE]
> this PR was generated automatically by ``batchpr`` :robot:

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [ ] added relevant milestone
- [ ] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by braingram** [comment]
Would you run regression tests with this change with numpy 1.x?

**Comment by zacharyburnett** [comment]
> Would you run regression tests with this change with numpy 1.x?

sure, here: https://plwishmaster.stsci.edu:8081/blue/organizations/jenkins/RT%2FJWST-Developers-Pull-Requests/detail/JWST-Developers-Pull-Requests/1494/pipeline

**Review by braingram — APPROVED** [review]
Thanks!

**Review by hbushouse — COMMENTED** [review]
So are we ready, willing, able to merge this now? We're OK with switching from numpy 1.x to 2.0?  I'm actually surprised that the regtests did NOT show any differences. I would've expected at least small numerical differences. What am I missing?

**Comment by stscirij** [comment]
Looks like the tests ran with numpy < 2.0

**Comment by braingram** [comment]
> So are we ready, willing, able to merge this now? We're OK with switching from numpy 1.x to 2.0? I'm actually surprised that the regtests did NOT show any differences. I would've expected at least small numerical differences. What am I missing?

The changes in this PR don't quite switch numpy to 2.0. As @stscirij noted the tests all ran with numpy 1.x (as expected) since many of the jwst dependencies do not yet have released versions that support numpy 2.0.

The main change in this PR is that this makes the jwst C extension compatible with both numpy 2.0 and numpy 1.x. This is related to the 5th item in the "Key guidance for users and downstream package authors" list here: https://github.com/numpy/numpy/issues/24300#issue-1828940995

We might want to consider making a patch release with `numpy<2` as I don't see this in the release branch:
https://github.com/spacetelescope/jwst/blob/f57f0b04cd93443ff36254ef389d98396544088f/pyproject.toml#L27
If this isn't pinned a user that installs the pipeline after the numpy 2 release may end up also installing numpy 2.0 and experience issues.

File: pyproject.toml [hunk 1]
```diff
@@ -24,7 +24,7 @@ dependencies = [
     "crds>=11.17.14",
     "drizzle>=1.15.0",
     "gwcs>=0.21.0,<0.22.0",
-    "numpy>=1.22",
+    "numpy>=1.22,<2.0",
```
Comment (braingram) [inline pyproject.toml:27]: Setting this pin on the development branch will means we can't test with numpy 2.0. Can we set this on the release branch?

**Comment by hbushouse** [comment]
> > So are we ready, willing, able to merge this now? We're OK with switching from numpy 1.x to 2.0? I'm actually surprised that the regtests did NOT show any differences. I would've expected at least small numerical differences. What am I missing?
> 
> The changes in this PR don't quite switch numpy to 2.0. As @stscirij noted the tests all ran with numpy 1.x (as expected) since many of the jwst dependencies do not yet have released versions that support numpy 2.0.
> 
> The main change in this PR is that this makes the jwst C extension compatible with both numpy 2.0 and numpy 1.x. This is related to the 5th item in the "Key guidance for users and downstream package authors" list here: [numpy/numpy#24300 (comment)](https://github.com/numpy/numpy/issues/24300#issue-1828940995)
> 
> We might want to consider making a patch release with `numpy<2` as I don't see this in the release branch:
> 
> https://github.com/spacetelescope/jwst/blob/f57f0b04cd93443ff36254ef389d98396544088f/pyproject.toml#L27
> 
> 
> If this isn't pinned a user that installs the pipeline after the numpy 2 release may end up also installing numpy 2.0 and experience issues.

OK, I was confused by the line here: https://github.com/spacetelescope/jwst/pull/8527/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R147
which looks to me like it's requiring numpy >= 2.0. But I guess that's just for certain kinds of builds or something?

**Comment by zacharyburnett** [comment]
> I'm not so sure for the stsci ones.

I'm tracking Numpy 2.0 builds and testing here: https://github.com/orgs/spacetelescope/projects/45/views/2

I've only added SCSB ones so far within this "calibration pipeline" sphere of dependencies

> * we've tested again with numpy 2.0 (to be sure the numpy pin in our install requirements isn't a lie)

I'll go through the devdeps testing and make sure

**Comment by hbushouse** [comment]
What's the status and intent of this? Are stakeholders happy enough to merge it?

**Comment by hbushouse** [comment]
In line https://github.com/spacetelescope/jwst/pull/8527/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R27 should we have an upper limit pin on numpy in order to avoid using 2.0 yet?

**Comment by braingram** [comment]
Oof, I see now that I edited https://github.com/spacetelescope/jwst/pull/8527#issuecomment-2153100591 instead of "Quote reply" Sorry about that @hbushouse

> In line https://github.com/spacetelescope/jwst/pull/8527/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R27 should we have an upper limit pin on numpy in order to avoid using 2.0 yet?

Pinning to "<2.0" in the release branch seems worth doing (as we haven't been able to test against numpy 2.0 due to dependency issues). Pinning in the main branch will make devdeps testing difficult.

I'm happy with the changes as they currently stand.

I just added the devdeps label so we can see the remaining failures (likely mostly from dependencies not having releases with numpy 2.0 support).

**Comment by braingram** [comment]
> I just added the devdeps label so we can see the remaining failures (likely mostly from dependencies not having releases with numpy 2.0 support).

The label didn't trigger the jobs but perhaps the resync when the changelog is updated will trigger this.

**Comment by braingram** [comment]
This could use `2.0` (instead of rc) when it's updated.

File: pyproject.toml [hunk 2]
```diff
@@ -144,7 +144,7 @@ requires = [
     "setuptools>=61.2",
     "setuptools_scm[toml]>=3.4",
     "wheel",
-    "oldest-supported-numpy",
+    "numpy>=2.0.0rc2",
```
Comment (zacharyburnett) [inline pyproject.toml:147]:
```suggestion
    "numpy>=2.0",
```

**Comment by braingram** [comment]
Any idea why the docs failed?

**Pull request merged**